extensions/socket/internal.m

```diff
@@ -0,0 +1,362 @@
+#import <LuaSkin/LuaSkin.h>
+#import <CocoaAsyncSocket/GCDAsyncSocket.h>
+
+static const char *USERDATA_TAG = "hs.socket";
+
+int refTable;
+
+@interface HSAsyncSocket : GCDAsyncSocket
+@property int callback;
+@end
+
+
+static void callback(HSAsyncSocket *asyncSocket, NSData *data) {
+    LuaSkin *skin = [LuaSkin shared];
+    NSString *utf8Data = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+
+    if (!asyncSocket.callback) {
+        lua_getglobal(skin.L, "print");
+    } else {
+        [skin pushLuaRef:refTable ref:asyncSocket.callback];
+    }
+
+    lua_pushstring(skin.L, [utf8Data UTF8String]);
+
+    if (![skin protectedCallAndTraceback:1 nresults:0]) {
+        const char *errorMsg = lua_tostring(skin.L, -1);
+        [skin logError:[NSString stringWithFormat:@"hs.socket callback error: %s", errorMsg]];
+    }
+}
+
+
+@implementation HSAsyncSocket
+
+- (id)init {
+    return [super initWithDelegate:self delegateQueue:dispatch_get_main_queue()];
+}
+
+- (void)socket:(HSAsyncSocket *)sock didConnectToHost:(NSString *)host port:(uint16_t)port {
+    [[LuaSkin shared] logInfo:@"Socket connected"];
+}
+
+- (void)socketDidDisconnect:(HSAsyncSocket *)sock withError:(NSError *)err {
+    [[LuaSkin shared] logInfo:@"Socket disconnected"];
+}
+
+- (void)socket:(HSAsyncSocket *)sock didWriteDataWithTag:(long)tag {
+    [[LuaSkin shared] logInfo:@"Data written to socket"];
+}
+
+- (void)socket:(HSAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag  {
+    [[LuaSkin shared] logInfo:@"Data read from socket"];
+
+    callback(self, data);
+}
+
+@end
+
+
+/// hs.socket.new(host, port) -> hs.socket object
+/// Constructor
+/// Creates an asynchronous TCP socket object for reading (with callbacks) and writing
+///
+/// Parameters:
+///  * host - A string containing the hostname or IP address
+///  * port - A port number [1024-65535]. Ports [1-1023] are privileged
+///  * fn - An optional callback function accepting a single parameter to process data. Can be set with `:setCallback`
+///
+/// Returns:
+///  * An `hs.socket` object
+///
+static int socket_new(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TSTRING, LS_TNUMBER, LS_TFUNCTION|LS_TNIL|LS_TOPTIONAL, LS_TBREAK];
+
+    HSAsyncSocket *asyncSocket = [[HSAsyncSocket alloc] init];
+
+    NSString *theHost = [skin toNSObjectAtIndex:1];
+    NSNumber *thePort = [skin toNSObjectAtIndex:2];
+    if (lua_type(L, 3) == LUA_TFUNCTION) {
+        lua_pushvalue(L, 3);
+        asyncSocket.callback = [skin luaRef:refTable];
+    }
+
+    NSError *err;
+    if (![asyncSocket connectToHost:theHost onPort:[thePort unsignedShortValue] error:&err]) {
+        [skin logError:[NSString stringWithFormat:@"Unable to connect: %@", err]];
+    }
+
+    [skin pushNSObject:asyncSocket];
```

Comment:
(I'd suggest copying the very simple struct userdata object pattern that I used in hs.chooser, IMO it's the easiest one to reason about)